### ISSUE TITLE
Can't connect to MongoDB when starting syslog-ng as a service

### ISSUE THREAD

**Issue by jenmontes** [balabit/syslog-ng#685]
I have an `apps.conf` file in `/etc/syslog-ng/conf.d/` with the following destination:

```
destination d_mongodb {
    mongodb();
};
```

If I start syslog-ng with `service syslog-ng start`, my `/var/log/messages` is spammed with `Error connecting to MongoDB; driver='d_mongodb#0'` and no logs are written. However, if I start it with `syslog-ng`, the logs are written to MongoDB as expected.

I have attempted to fix this problem by setting the `servers` destination option to both `"127.0.0.1:27017"` and `"localhost:27017"`. I have also attempted to use `path("/tmp/mongodb-27017.sock")` instead. All of these options worked when calling `syslog-ng` but did not work when starting the service.

System stats:
  * syslog-ng version 3.7.1
  * MongoDB version 2.6.11
  * CentOS release 6.7 (Final)
  * Kernel: 2.6.32-573.3.1.el6.x86_64

**Comment by faxm0dem** [comment]
Where did you get the package from?

**Comment by jenmontes** [comment]
I just followed the instructions on the Balabit blog: https://czanik.blogs.balabit.com/2015/09/installing-syslog-ng-ose-3-7-1-on-rhel6-and-centos6/

The repo file can be found here: http://copr.fedoraproject.org/coprs/czanik/syslog-ng37epel6/

Besides syslog-ng, I also installed syslog-ng-json and syslog-ng-mongodb.

**Comment by faxm0dem** [comment]
I don't have a test system at reach. Could you paste the output of `sh -x /etc/init.d/syslog-ng start` please?

**Comment by jenmontes** [comment]
Stdout:

```
syslog-ng is stopped
Starting syslog-ng: [60G[[0;32m  OK  [0;39m]
```

Stderr:

```
+ . /etc/init.d/functions
++ TEXTDOMAIN=initscripts
++ umask 022
++ PATH=/sbin:/usr/sbin:/bin:/usr/bin
++ export PATH
++ '[' -z '' ']'
++ COLUMNS=80
++ '[' -z '' ']'
+++ /sbin/consoletype
++ CONSOLETYPE=pty
++ '[' -f /etc/sysconfig/i18n -a -z '' -a -z '' ']'
++ . /etc/profile.d/lang.sh
++ unset LANGSH_SOURCED
++ '[' -z '' ']'
++ '[' -f /etc/sysconfig/init ']'
++ . /etc/sysconfig/init
+++ BOOTUP=color
+++ RES_COL=60
+++ MOVE_TO_COL='echo -en \033[60G'
+++ SETCOLOR_SUCCESS='echo -en \033[0;32m'
+++ SETCOLOR_FAILURE='echo -en \033[0;31m'
+++ SETCOLOR_WARNING='echo -en \033[0;33m'
+++ SETCOLOR_NORMAL='echo -en \033[0;39m'
+++ PROMPT=yes
+++ AUTOSWAP=no
+++ ACTIVE_CONSOLES='/dev/tty[1-6]'
+++ SINGLE=/sbin/sushell
++ '[' pty = serial ']'
++ __sed_discard_ignored_files='/\(~\|\.bak\|\.orig\|\.rpmnew\|\.rpmorig\|\.rpmsave\)$/d'
+++ cat /proc/cmdline
++ strstr 'ro root=/dev/mapper/vg_fcpdb-lv_root rd_NO_LUKS LANG=en_US.UTF-8 rd_NO_MD SYSFONT=latarcyrheb-sun16 crashkernel=129M@0M rd_LVM_LV=vg_fcpdb/lv_root  KEYBOARDTYPE=pc KEYTABLE=us rd_LVM_LV=vg_fcpdb/lv_swap rd_NO_DM rhgb quiet' rc.debug
++ '[' 'ro root=/dev/mapper/vg_fcpdb-lv_root rd_NO_LUKS LANG=en_US.UTF-8 rd_NO_MD SYSFONT=latarcyrheb-sun16 crashkernel=129M@0M rd_LVM_LV=vg_fcpdb/lv_root  KEYBOARDTYPE=pc KEYTABLE=us rd_LVM_LV=vg_fcpdb/lv_swap rd_NO_DM rhgb quiet' = 'ro root=/dev/mapper/vg_fcpdb-lv_root rd_NO_LUKS LANG=en_US.UTF-8 rd_NO_MD SYSFONT=latarcyrheb-sun16 crashkernel=129M@0M rd_LVM_LV=vg_fcpdb/lv_root  KEYBOARDTYPE=pc KEYTABLE=us rd_LVM_LV=vg_fcpdb/lv_swap rd_NO_DM rhgb quiet' ']'
++ return 1
+ SYSLOGNG_OPTIONS='-p /var/run/syslog-ng.pid'
+ '[' -e /etc/sysconfig/syslog-ng ']'
+ . /etc/sysconfig/syslog-ng
++ SYSLOGNG_OPTIONS='-p /var/run/syslog-ng.pid'
+ RETVAL=0
+ case "$1" in
+ start
+ verify_config
+ check_syntax
+ '[' -x /sbin/syslog-ng ']'
+ syslog-ng -s -p /var/run/syslog-ng.pid
+ RETVAL=0
+ return 0
+ RETVAL=0
+ '[' 0 -eq 0 ']'
+ status syslog-ng
+ local base pid lock_file= pid_file= binary=
+ '[' 1 = 0 ']'
+ '[' syslog-ng = -p ']'
+ '[' syslog-ng = -l ']'
+ '[' syslog-ng = -b ']'
+ base=syslog-ng
+ __pids_var_run syslog-ng '' ''
+ local base=syslog-ng
+ local pid_file=/var/run/syslog-ng.pid
++ /usr/bin/dirname /var/run/syslog-ng.pid
+ local pid_dir=/var/run
+ local binary=
+ '[' -d /var/run -a '!' -r /var/run ']'
+ pid=
+ '[' -f /var/run/syslog-ng.pid ']'
+ return 3
+ RC=3
+ '[' -z '' -a -z '' ']'
++ __pids_pidof syslog-ng
++ pidof -c -m -o 3325 -o 3215 -o %PPID -x syslog-ng
++ pidof -c -m -o 3325 -o 3215 -o %PPID -x syslog-ng
+ pid=
+ '[' -n '' ']'
+ case "$RC" in
+ '[' -z '' ']'
+ lock_file=syslog-ng
+ '[' -f /var/lock/subsys/syslog-ng ']'
+ echo 'syslog-ng is stopped'
+ return 3
+ echo -n 'Starting syslog-ng: '
+ daemon syslog-ng -p /var/run/syslog-ng.pid
+ local gotbase= force= nicelevel corelimit
+ local pid base= user= nice= bg= pid_file=
+ local cgroup=
+ nicelevel=0
+ '[' syslog-ng '!=' syslog-ng ']'
+ '[' -z '' ']'
+ base=syslog-ng
+ __pids_var_run syslog-ng ''
+ local base=syslog-ng
+ local pid_file=/var/run/syslog-ng.pid
++ /usr/bin/dirname /var/run/syslog-ng.pid
+ local pid_dir=/var/run
+ local binary=
+ '[' -d /var/run -a '!' -r /var/run ']'
+ pid=
+ '[' -f /var/run/syslog-ng.pid ']'
+ return 3
+ '[' -n '' -a -z '' ']'
+ corelimit='ulimit -S -c 0'
+ '[' -n '' ']'
+ '[' -n '' ']'
+ '[' color = verbose -a -z '' ']'
+ '[' -z '' ']'
+ /bin/bash -c 'ulimit -S -c 0 >/dev/null 2>&1 ; syslog-ng -p /var/run/syslog-ng.pid'
+ '[' 0 -eq 0 ']'
+ success 'syslog-ng startup'
+ '[' color '!=' verbose -a -z '' ']'
+ echo_success
+ '[' color = color ']'
+ echo -en '\033[60G'
+ echo -n '['
+ '[' color = color ']'
+ echo -en '\033[0;32m'
+ echo -n '  OK  '
+ '[' color = color ']'
+ echo -en '\033[0;39m'
+ echo -n ']'
+ echo -ne '\r'
+ return 0
+ return 0
+ RETVAL=0
+ echo
+ '[' 0 -eq 0 ']'
+ touch /var/lock/subsys/syslog-ng
+ return 0
+ exit 0
```

The strange thing is that if I start the service with `sh`, MongoDB logs are written as expected, but starting the service normally still doesn't work!

**Comment by mitzkia** [comment]
Hi, 
I setup the same environment, but can not reproduce your issue. In my case syslog-ng (started with `service syslog-ng start`) can connect to mongodb server, which is listening on the same host.
Can you paste here other syslog-ng relevant log lines from /var/log/messages (when syslog-ng can not connect to mongodb)? For example: 
`Initializing MongoDB destination; address='127.0.0.1', port='27017', database='syslog', collection='messages', driver='d_mongo#0'`
Can you verify that when syslog-ng starts mongodb server listens on 127.0.0.1:27017?

**Comment by ihrwein** [comment]
Do you use SELinux? Maybe your system is in `targeted` mode and if syslog-ng is started by the init it gets confined and blocked to connect to a non-syslog port. If you run it manually  syslog-ng can run unconstrained.

I'd check the output of the following commands:
```
sestatus
```

And your audit log would be also interesting:
```
sealert -a /var/log/audit/audit.log
``

**Comment by ihrwein** [comment]
@jenmontes Do you have an update on this issue?

**Comment by jenmontes** [comment]
Yes! I apologize for not reporting back sooner.

After starting from a completely clean install of CentOS 6.7 I can't reproduce my original issue. There must have been something in my original installation that was interfering somehow.

To facilitate testing, I created a gist with a Vagrantfile and a bootstrap script that installs everything needed to get Syslog NG working on CentOS 6.7 with MongoDB: https://gist.github.com/jenmontes/c941145200f736471448

Once the environment is ready, I can SSH into it and run the `mongo` shell with the following commands:

```
use syslog
db.messages.find()
```

and the latest log entries are shown, as expected.

Closing this issue.
